File: 2022/src/Day03.scala
object Day03 {
  def main(args: Array[String]): Unit = {
    val lines = util.Using.resource(io.Source.fromFile("2022/input/03.txt")) { _
      .getLines
      .map(line => line.map(priority).toList)
      .toList
    }

    val part1 = lines
      .map(line => {
        val first = line.slice(0, line.length / 2).toSet
        val second = line.slice(line.length / 2, line.length).toSet
        first.intersect(second).last
      })
      .sum
    println(s"Part 1: $part1")

    val part2 = Range(0, lines.length, 3)
      .map(idx => {
        lines(idx)
          .toSet
          .intersect(lines(idx + 1).toSet)
          .intersect(lines(idx + 2).toSet)
          .last
      })
      .sum
    println(s"Part 2: $part2")
  }

  private def priority(c: Char): Int = {
    if ('a' <= c && c <= 'z')
      c - 'a' + 1
    else
      c - 'A' + 27
  }
}

File: aoc2024/src/main/kotlin/day03/Day03.kt
package day03

import java.io.File

fun main() {
    val input = File("input/03.txt").readText()
    println("Part 1: ${part1(input)}")
    println("Part 2: ${part2(input)}")
}

fun part1(input: String): Int = findAll(input).sumOf { it.first * it.second }
fun part2(input: String): Int = findAllConditional(input).sumOf { it.first * it.second }

fun findAll(input: String): List<Pair<Int,Int>> {
    val pattern = """mul\((\d+),(\d+)\)""".toRegex()
    val result = mutableListOf<Pair<Int,Int>>()
    val matches = pattern.findAll(input)
    matches.forEach {
        result.add(Pair(it.groupValues[1].toInt(), it.groupValues[2].toInt()))
    }
    return result
}

fun findAllConditional(input: String): List<Pair<Int,Int>> {
    val pattern = """(do\(\)|don't\(\))""".toRegex()
    val matches = pattern.findAll(input)
    var updatedInput = ""
    var idx = 0
    var enabled = true
    matches.forEach {
        if (enabled) {
            updatedInput += input.substring(idx..<it.range.first)
        }
        when (it.value) {
            "do()" -> enabled = true
            "don't()" -> enabled = false
        }
        idx = it.range.last + 1
    }
    if (enabled) {
        updatedInput += input.substring(idx..<input.length)
    }
    return findAll(updatedInput)
}
